keras_cv/metrics/coco/coco_pymetric.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import six
+import tensorflow as tf
+from keras.callbacks import Callback
+
+from keras_cv import bounding_box
+from keras_cv.metrics.coco.pycoco_utils import compute_pycoco_metrics
+
+
+class COCOPyMetric(Callback):
+    def __init__(self, bounding_box_format):
+        self.model = None
+        self.bounding_box_format = bounding_box_format
+        super().__init__()
+
+    def set_model(self, model):
+        self.model = model
+
+    def on_epoch_end(self, epoch, logs=None):
+        logs = logs or {}
+
+        dataset = self.model._eval_data_handler._dataset
+        gt = {}
+        preds = {}
+
+        for i, batch in enumerate(dataset):
+            gt_i, preds_i = self._eval_batch(batch, i)
+
+            for k, v in six.iteritems(preds_i):
+                if k not in preds:
+                    preds[k] = [v]
+                else:
+                    preds[k].append(v)
+
+            for k, v in six.iteritems(gt_i):
+                if k not in gt:
+                    gt[k] = [v]
+                else:
+                    gt[k].append(v)
+
+        metrics = compute_pycoco_metrics(gt, preds)
+        # Mark these as validation metrics by prepending a val_ prefix
+        metrics = {"val_" + name: val for name, val in metrics.items()}
+
+        logs.update(metrics)
+
+    def _eval_batch(self, batch, index):
+        images, gt_boxes = batch
+        batch_size = images.shape[0]
+        height = images.shape[1]
+        width = images.shape[2]
+
+        gt_boxes = bounding_box.convert_format(
+            gt_boxes, source=self.bounding_box_format, target="yxyx"
+        )
+
+        gt_classes = gt_boxes[:, :, 4]
+        gt_boxes = gt_boxes[:, :, :4]
+
+        source_ids = tf.strings.join(
```

Comment:
I'd recommend taking the source id from the input dataset, this ideally should just be from "00032.jpg" -> 32

Reply:
IMO this is a bad user experience, because it requires that the dataset has the source ID as a field. This isn't really exposed to the user in any meaningful way, it's just used internally by pycocoutils as a key for a row in the dataset.

If we require "source_id" as a field in the dataset, then users will have to manufacture a source ID themselves for datasets that don't already have one

Reply:
It is part of tfds, correct? https://www.tensorflow.org/datasets/catalog/coco
So in the future, we will probably provide some tools for users to find out bad predictions based on the metric output, it'd be nice to associate image id / filename with bad predictions

Reply:
Yes, this is part of TFDS for COCO. However, using it here:
- requires users to parse that field and add it to the train/val datasets
- adds a lot of friction for using these metrics with any other dataset

Is there any value add to using it? From what Ic an see, we get the exact same behavior with a simpler API by leaving it out.

Reply:
Ok, that makes sense